### PR TITLE
Add support for overriding IdP Entity ID per SAML SP

### PR DESCRIPTION
**Pull request by DinikaSen**
## Proposed changes in this pull request
These modifications, 
 - add support for overriding Identity Provider Entity ID per Service Provider. - A value for "IdP Entity ID Alias" can be configured within the Service Provider's SAML SSO inbound authentication configuration, and then it will override the "Identity Provider Entity ID" value specified in SAML SSO inbound authentication configuration in the Resident IdP.
- add support for configuring multiple SAML SSO inbound authentication configurations for a single SAML requset issuer with the use of an issuer qualifier value.

NOTE: This PR depends on the changes done to carbon-identity-framework in PR - wso2/carbon-identity-framework/pull/1999. Therefore, the Travis CI build fails until the other PR is merged.

### When should this PR be merged
This PR should be merged after merging the following PR -
wso2/carbon-identity-framework/pull/1999

#### Tests
Unit tests - Implemented unit tests for newly added functions in SAMLSSOUtil class

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.saml.ui/src/main/resources/org/wso2/carbon/identity/sso/saml/ui/i18n/Resources.properties [hunk 1]
```diff
@@ -114,3 +117,5 @@ sp.saml.metadata.certificate.warn=Since this the super tenant, Keystore is manag
   uploading this file. Also, make sure you have replicated it among other nodes as well. Do you want to continue?
 sp.enable.saml2.artifact.binding=Enable SAML2 Artifact Binding
 sp.enable.signature.validation.artifact.resolve=Enable Signature Validation in Artifact Resolve Request
+enable.idp.entity.id.alias.help=Default value is the IdP Entity ID value specified in Resident IdP
+idp.entity.id.alias=IdP Entity ID Alias
```
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.sso.saml.ui/src/main/resources/org/wso2/carbon/identity/sso/saml/ui/i18n/Resources.properties]: add new line

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java [hunk 2]
```diff
@@ -265,7 +273,14 @@ private SAMLSSOServiceProviderDTO createSAMLSSOServiceProviderDTO(SAMLSSOService
             throw IdentityException.error(message);
         }
 
+        if (serviceProviderDO.getIssuerQualifier() != null && serviceProviderDO.getIssuerQualifier().contains("@")) {
```
Comment (ayshsandu) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/admin/SAMLSSOConfigAdmin.java]: check isNotBlank

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java [hunk 3]
```diff
@@ -1662,6 +1725,68 @@ public static boolean isSAMLIssuerExists(String issuerName, String tenantDomain)
...
+     * @param issuerWithQualifier Issuer value saved in the registry.
+     * @param tenantDomain
+     * @return true, if a SAML SP exists in the registry with the issuer value similar to "issuerWithQualifier"
+     * and the SAML request's issuer is equal to the issuer without qualifier.
+     */
+    public static boolean isValidSAMLIssuer(String issuerName, String issuerWithQualifier, String tenantDomain)
+            throws IdentitySAML2SSOException {
+
+        if (isSAMLIssuerExists(issuerWithQualifier, tenantDomain)) {
+            SAMLSSOServiceProviderDO sp = getSAMLSSOServiceProvider(issuerWithQualifier, tenantDomain);
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java]: Why we need to load SAML service provider here? can't we use persistenceManager.isServiceProviderExists(registry, issuerName)?

File: components/org.wso2.carbon.identity.sso.saml/src/test/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtilTest.java [hunk 4]
```diff
@@ -109,6 +119,23 @@ private void prepareForGetIssuer() throws Exception {
         when(identityProviderManager.getResidentIdP(anyString())).thenReturn(identityProvider);
     }
 
+    private void prepareForGetSAMLSSOServiceProvider() throws Exception {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/test/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtilTest.java]: add an empty line

File: components/org.wso2.carbon.identity.sso.saml/src/test/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtilTest.java [hunk 5]
```diff
@@ -329,4 +356,27 @@ public void testBuildResponseStatus() {
         assertEquals(status.getStatusMessage().getMessage(), statusMsg, "Status Message is not properly set in " +
                 "the Status object.");
     }
+
+    @Test
+    public void testIsValidSAMLIssuer() throws Exception {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/test/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtilTest.java]: add an empty line

File: components/org.wso2.carbon.identity.sso.saml/src/test/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtilTest.java [hunk 6]
```diff
@@ -329,4 +356,27 @@ public void testBuildResponseStatus() {
...
+        SAMLSSOUtil.setRealmService(realmService);
+        prepareForGetSAMLSSOServiceProvider();
+        TestUtils.startTenantFlow(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+        boolean isSAMLIssuerValidated = SAMLSSOUtil.isValidSAMLIssuer(TestConstants.TRAVELOCITY_ISSUER,
+                TestConstants.ISSUER_WITH_QUALIFIER, MultitenantConstants.SUPER_TENANT_DOMAIN_NAME);
+        assertTrue(isSAMLIssuerValidated, "SAML Issuer should be validated");
+    }
+
+    @Test
+    public void testGetIssuerWhenEntityIDAliasEnabled() throws Exception {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.sso.saml/src/test/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtilTest.java]: add an empty line

**Comment by malithie** [comment]
@ayshsandu , as this is with respect to a new feature, can we create a feature branch and get this PR merged

**Comment by ayshsandu** [comment]
This same code is merged in https://github.com/wso2-extensions/identity-inbound-auth-saml/pull/243. Hence, closing the PR.

File: components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl [hunk 7]
```diff
@@ -847,4 +849,4 @@
             <http:address location="https://localhost:9443/services/IdentitySAMLSSOConfigService.IdentitySAMLSSOConfigServiceHttpsEndpoint/"/>
         </wsdl:port>
     </wsdl:service>
-</wsdl:definitions>
+</wsdl:definitions>
```
Comment (Kanapriya) [inline components/org.wso2.carbon.identity.sso.saml.stub/src/main/resources/IdentitySAMLSSOConfigService.wsdl:852]: Add new line here

**Pull request closed without merging**